src/test/csf/TrustGraph.h

```diff
@@ -0,0 +1,176 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012-2017 Ripple Labs Inc
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_TEST_CSF_UNL_H_INCLUDED
+#define RIPPLE_TEST_CSF_UNL_H_INCLUDED
+
+#include <test/csf/random.h>
+#include <boost/container/flat_set.hpp>
+#include <boost/optional.hpp>
+#include <chrono>
+#include <numeric>
+#include <random>
+#include <vector>
+
+namespace ripple {
+namespace test {
+namespace csf {
+
+/** Trust graph
+
+    Trust is a directed relationship from a node i to node j.
+    If node i trusts node j, then node i has node j in its UNL.
+    This class wraps a digraph representing the trust relationships for all peers
+    in the simulation.
+*/
+template <class Peer>
+class TrustGraph
+{
+    using Graph = Digraph<Peer>;
+
+    Graph graph_;
+
+public:
+    /** Create an empty trust graph
+     */
+    TrustGraph() = default;
+
+    Graph const&
+    graph()
+    {
+        return graph_;
+    }
+
+    /** Create trust
+
+        Establish trust between Peer `from` and Peer `to`; as if `from` put `to`
+        in its UNL.
+
+        @param from The peer granting trust
+        @param to The peer receiving trust
+
+    */
+    void
+    trust(Peer const& from, Peer const& to)
```

Comment:
Both `Peer` and `PeerGroup` provide the asymmetric interface you prefer.  Most simulations/tests use those .  But since the `TrustGraph` is owned by the simulation at large, so doesn't have a way to break symmetry when building the connection. Although clients can interact with `TrustGraph` directly, since it isn't the common way I prefer having the symmetric interface.